### PR TITLE
FOUR-8231

### PR DESCRIPTION
**Pull request by luNunezProcessmaker**
## Issue & Reproduction Steps
N/A

## Solution
N/A

## How to Test
- Go to admin
- Select edit an user
- Go to groups tab
- Change the groups and SAVE
- Check the security_logs table for a new log "groupMembershipChanged"

## Related Tickets & Packages
https://processmaker.atlassian.net/browse/FOUR-8231

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

ci:deploy

### CODE REVIEW

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

File: ProcessMaker/Events/GroupUsersUpdated.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                    ],
+                    '- user' => [
+                        'link' => route('users.edit', $user),
+                        'label' => $user->username
+                    ]
+                ];
+                break;
+        }
+    }
+
```
Comment (pmPaulis) [inline ProcessMaker/Events/GroupUsersUpdated.php:75]: Add description

File: ProcessMaker/Events/GroupUsersUpdated.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                ];
+                break;
+        }
+    }
+    
+    public function getData(): array
+    {
+        return $this->data;
+    }
+
```
Comment (pmPaulis) [inline ProcessMaker/Events/GroupUsersUpdated.php:80]: Add description

File: ProcessMaker/Events/UserGroupMembershipUpdated.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        $this->userUpdated = $userUpdated;
+        $this->buildData($data);
+    }
+
+    /**
+     * Building the data
+     */
+    public function buildData($data) {
+        
+        $groups_deleted = [];
```
Comment (pmPaulis) [inline ProcessMaker/Events/UserGroupMembershipUpdated.php:44]: Lower camel case groupsDeleted

File: ProcessMaker/Events/UserGroupMembershipUpdated.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        $groups_added = [];
```
Comment (pmPaulis) [inline ProcessMaker/Events/UserGroupMembershipUpdated.php:45]: Lower camel case groupsAdded

File: ProcessMaker/Events/UserGroupMembershipUpdated.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace ProcessMaker\Events;
+
+use Illuminate\Broadcasting\Channel;
+use Illuminate\Broadcasting\InteractsWithSockets;
```
Comment (pmPaulis) [inline ProcessMaker/Events/UserGroupMembershipUpdated.php:6]: This is declared but is not used, check for all

File: ProcessMaker/Events/GroupUsersUpdated.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace ProcessMaker\Events;
+
+use Illuminate\Broadcasting\Channel;
+use Illuminate\Broadcasting\InteractsWithSockets;
```
Comment (pmPaulis) [inline ProcessMaker/Events/GroupUsersUpdated.php:6]: This is declared but is not used, check for all

File: ProcessMaker/Events/UserGroupMembershipUpdated.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    public $data;
+
+    /**
+     * Create a new event instance.
+     *
+     * @return void
+     */
+    public function __construct(Array $data, User $userUpdated)
+    {
+        $this->user = Auth::user();
```
Comment (pmPaulis) [inline ProcessMaker/Events/UserGroupMembershipUpdated.php:34]: Remove this line $this->user = Auth::user();

**Review by pmPaulis — CHANGES_REQUESTED** [review]

File: ProcessMaker/Http/Controllers/Api/UserController.php [hunk 8]
```diff
@@ -309,14 +310,14 @@ public function updateGroups(User $user, Request $request)
...
-                $user->groups()->sync($groups);
+                $data = $user->groups()->sync($groups);
             } else {
                 $user->groups()->detach();
             }
         } else {
             return response([], 400);
         }
-
+        event(new UserGroupMembershipUpdated($data, $user));
```
Comment (paulosakamoto) [inline ProcessMaker/Http/Controllers/Api/UserController.php:320]: When there are no groups and the groups are detached we get into a bug situation as `$data` is undefined.

File: ProcessMaker/Events/UserGroupMembershipUpdated.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+    public $userUpdated;
+    public $data;
+
+    /**
+     * Create a new event instance.
+     *
+     * @return void
+     */
+    public function __construct(Array $data, User $userUpdated)
```
Comment (paulosakamoto) [inline ProcessMaker/Events/UserGroupMembershipUpdated.php:22]: Could you please change `Array` to lowercase?

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

File: ProcessMaker/Http/Controllers/Api/GroupMemberController.php [hunk 10]
```diff
@@ -134,6 +135,8 @@ public function store(Request $request)
         $group_member->member()->associate($member);
         $group_member->saveOrFail();
 
+        event(new GroupUsersUpdated($group->id , $request->input('member_id'), 'added', $request->input('member_type')));
```
Comment (pmPaulis) [inline ProcessMaker/Http/Controllers/Api/GroupMemberController.php:138]: Avoid this hardcoded value 'added'
`GroupUsersUpdated::ADDED`

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

**Review by pmPaulis — CHANGES_REQUESTED** [review]

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

File: ProcessMaker/Providers/EventServiceProvider.php [hunk 11]
```diff
@@ -50,6 +50,12 @@ class EventServiceProvider extends ServiceProvider
         ],
         'ProcessMaker\Events\SettingsUpdated' => [
             'ProcessMaker\Listeners\SecurityLogger',
+        ], 
+        'ProcessMaker\Events\UserGroupMembershipUpdated' => [
```
Comment (pmPaulis) [inline ProcessMaker/Providers/EventServiceProvider.php]: Reorder UserGroupMembershipUpdated after UserDeleted

File: ProcessMaker/Providers/EventServiceProvider.php [hunk 12]
```diff
@@ -50,6 +50,12 @@ class EventServiceProvider extends ServiceProvider
...
+            'ProcessMaker\Listeners\SecurityLogger',
+        ],
+        'ProcessMaker\Events\GroupUsersUpdated' => [
```
Comment (pmPaulis) [inline ProcessMaker/Providers/EventServiceProvider.php]: Reorder GroupUsersUpdated after GroupDeleted

**Review by pmPaulis — APPROVED** [review]

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

**Comment by pmkerberos** [comment]
QA server was successfully deployed https://ci-3895896475.eng.processmaker.net

**Pull request merged**